Add stream filter to Recent topics.

amanagr opened the pull request:
![RT-stream_filter](https://user-images.githubusercontent.com/25124304/141824633-4599e8b2-fe6a-4971-8d42-4f43378658a7.gif)

timabbott commented:
I'm excited to see this getting attention!  @aryanshridhar FYI, since you were working on making a multi-select version of the dropdown widget that we'd eventually hope to use here.

amanagr commented:
![image](https://user-images.githubusercontent.com/25124304/151554230-7d214b69-e2a7-4642-af68-5c229d59d220.png)

Super simple iteration of stream filter. Next iteration would be multi-select version of this. Then, add a function in backend to get last message of all topics in the stream and display them.

timabbott commented:
I think the top option should be "Clear filter" if you've already selected a stream.

![image](https://user-images.githubusercontent.com/2746074/155821170-0dfc68ea-e327-4a79-bbac-3c98a3da02e8.png)

The other thing is I think we should probably decorate the stream name with its icon, perhaps using the same styling we use in the "Move topic" stream selection UI; as it is you just have a "Denmark" and it's not obvious what role that word has.

![image](https://user-images.githubusercontent.com/2746074/155821194-c33ecc3b-1122-46aa-81fa-c5a94ae3d9ed.png)

alya commented:
Here are some thoughts from playing with the PR.

1. I agree that we should remove the "Filter streams" option, but maybe we should just have an "x" rather than "Clear filter"? I don't think we use that text elsewhere.
2. I agree that the pill is currently confusing once you have a stream selected, and it's worth trying adding stream decorators. I'm not sure that'll be sufficient, though, for making it not confusing... we can see.
3. I found it distracting to have the pill resize each time I changed the selection, moving everything else around in the process. I wonder if we should leave some extra space around the word so that it doesn't have to adjust all the time?
4. Long stream names are not handled well. Turns out we have the same issue in the "move message" UI, so let me know if you'd like me to file a separate issue for this. I guess we should just make the drop-down wider if needed in both places, like we do in the typeahead in the compose box, for example.
<img width="409" alt="Screen Shot 2022-03-01 at 1 28 29 PM" src="https://user-images.githubusercontent.com/2090066/156253191-0ac51cea-e935-4871-9b6b-fc3889c7186b.png">

timabbott commented:
I think this is one of the things I'll want to prioritize integrating once 5.0 is out the door.

alya commented:
@amanagr would you be up for updating this PR based on the feedback above?

amanagr commented:
Thanks for the reminder.

timabbott commented:
@amanagr just wanted to bump this for your TODO list; I think once we've done https://github.com/zulip/zulip/pull/20256#issuecomment-1055886341, we can do a chat.zulip.org deployment on this. We may very well find that actually merging it is blocked by data plumbing issues around scrolling to the bottom, but it might be pretty valuable for figuring out what those need to be to run it for a week.

alya commented:
Bumping this one again for @amanagr, in case that's helpful.

amanagr committed:
dropdown_list_widget: Allow caller to set its own render item template.

To be used by recent topics stream filter widget.

amanagr committed:
stream_events: Rerender recent topics on stream privacy change.

This should be done to avoid any possible data mismatch after
privacy change of a stream.

amanagr commented:
<img width="561" alt="Screenshot 2022-09-01 at 7 46 06 PM" src="https://user-images.githubusercontent.com/25124304/187937415-36883bd4-6f5f-418a-9d89-21dac4be1871.png">
<img width="561" alt="Screenshot 2022-09-01 at 7 46 33 PM" src="https://user-images.githubusercontent.com/25124304/187937424-ea328bfb-bf8b-4b4f-b987-fab400a87fce.png">
<img width="561" alt="Screenshot 2022-09-01 at 7 46 23 PM" src="https://user-images.githubusercontent.com/25124304/187937432-491b5287-7414-4665-9b32-44908b71642f.png">
<img width="561" alt="Screenshot 2022-09-01 at 7 46 11 PM" src="https://user-images.githubusercontent.com/25124304/187937442-e41f4892-27f5-47e0-a48b-d7b5df75219f.png">

@alya updated based on your feedback!

amanagr commented on static/templates/recent_topics_filters.hbs:
```diff
@@ -23,3 +23,6 @@
     {{/if}}
     {{t 'Participated' }}
 </button>
```
Move the stream filter widget before input box so that the change in the width of the button when selected stream is changed will be consumed by the input box and not move other filters around.

amanagr commented on static/templates/recent_topics_stream_selector.hbs:
```diff
@@ -0,0 +1,39 @@
+                {{> stream_privacy }}
+            </span>
+            <span id="{{widget_name}}_name"></span>
+            {{#if stream_id}}
+                <i class="fa fa-chevron-down"></i>
+            {{else}}
+                <i class="fa fa-times"></i>
+            {{/if}}
```
Added cross button when a stream is selected.

amanagr commented on static/styles/popovers.css:
```diff
@@ -660,6 +660,20 @@ ul {
 }
 
+.dropdown-menu {
+    @media (width > $md_min) {
+        min-width: 300px;
+
+        .dropdown-search {
+            text-align: center;
+
+            input {
+                width: 260px;
+            }
+        }
+    }
```
(I should move this to a separate commit.)

amanagr committed:
dropdown_list_widget: Increase width of the list.

This widget is mostly used for streams and since streams can have
long names, it makes sense for this to be a little more wider.

We only do it for wide screens.

amanagr committed:
recent_topics: Add button to filter streams.

amanagr commented:
Let me push another update then this should be ready for review.

amanagr commented:
<img width="561" alt="Screenshot 2022-09-02 at 1 29 26 AM" src="https://user-images.githubusercontent.com/25124304/188001446-6ab79ad6-33c2-4567-831f-5435c5ca0243.png">
screenshot with the cross icon.

alya commented:
Hmm, dark theme looks broken:

![Screen Shot 2022-09-01 at 2 59 06 PM](https://user-images.githubusercontent.com/2090066/188020430-f84b848a-949d-41be-97f8-7b80f5d885a7.png)

alya commented:
The other thing I noticed is that the term "All", which was already a bit of a misnomer regarding muted streams, becomes even more incorrect once this filter is added.

After chatting with @timabbott , let's just try deleting the "All" button. We can also consider adding a "reset" icon in the future if we feel the need (e.g. something along these lines).

![Screen Shot 2022-09-01 at 3 11 06 PM](https://user-images.githubusercontent.com/2090066/188021737-6384e883-1287-450e-a60b-0414cef170f9.png)

amanagr commented:
Need a reimplementation.
